### PR TITLE
[SPARK-49014][BUILD] Bump Apache Avro to 1.12.0

### PR DESCRIPTION
**Pull request by Fokko**
### What changes were proposed in this pull request?

This PR aims to update `Apache Avro` to 1.12.0.

### Why are the changes needed?

Apache Avro 1.12.0 is the latest feature release.
- https://github.com/apache/avro/releases/tag/release-1.12.0

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass the CIs.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -359,6 +359,11 @@
         <enabled>false</enabled>
       </snapshots>
     </repository>
+    <repository>
+      <id>avro-release-candidate</id>
+      <name>Avro Release Candidate</name>
+      <url>https://repository.apache.org/content/repositories/orgapacheavro-1037/</url>
```
Comment (dongjoon-hyun) [inline pom.xml:365]: According to the Apache Avro dev mailing list, this seems to be the correct RC1 location. Could you trigger once more if you want to test Avro 1.12.0 RC, @Fokko ?
- https://lists.apache.org/thread/tqj0zk7qqsgr6tw10247grz3dv3svhtn

Comment (dongjoon-hyun) [reply pom.xml:365]: To make CI happy, you need to update SBT file too.

https://github.com/apache/spark/blob/112a52d58816a842b4a2aa9f37367e1317623625/project/SparkBuild.scala#L272-L279

Comment (Fokko) [reply pom.xml:365]: Thanks @dongjoon-hyun for jumping in here. I wanted to test the RC with Spark to see if there are any breaking changes. I had it on my list to figure out how to fix SBT, but thanks for the pointer, that's helpful 👍

**Comment by dongjoon-hyun** [comment]
All tests passed. Thank you for updating. :)

**Comment by Fokko** [comment]
Thanks :) Looks good that all the tests pass on the [first RC](https://lists.apache.org/thread/f74q8gwc82khlzg4tp63lym3oqmofq05). Fingers crossed 🤞

**Comment by Fokko** [comment]
@dongjoon-hyun And it has been published :)

**Comment by dongjoon-hyun** [comment]
Thank you so much, @Fokko !

**Comment by dongjoon-hyun** [comment]
BTW, is this announced in Apache Avro website or GitHub repo? I cannot find this release yet.
- https://github.com/apache/avro/releases/
- https://avro.apache.org/project/download/

**Comment by Fokko** [comment]
@dongjoon-hyun We're in the process of releasing which is quite an effort since we do all the languages simultaneously. When all the convenience binaries are published, the announcement email will go out.

**Review by dongjoon-hyun — COMMENTED** [review]
No problem. Thank you. Then, let's wait until the official release because this is not a maintenance release. It's a new minor version upgrade.

**Review by dongjoon-hyun — APPROVED** [review]
I saw the newly created link. It's enough for us. 
- https://github.com/apache/avro/releases/tag/release-1.12.0

I revised the PR title. Now, +1, LGTM.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0-preview2.

**Comment by dongjoon-hyun** [comment]
Thank you again, @Fokko .

**Pull request closed without merging**